solana/pyth2wormhole/client/src/main.rs

```diff
@@ -190,301 +225,276 @@ fn handle_attest(
+    } = rlmtx.lock().await.clone();
+    RpcClient::new_with_timeout_and_commitment(url, timeout, commitment)
+}
 
-        let successes_pct = tx_successes as f64 / total * 100.0;
-        let send_failures_pct = tx_send_failures as f64 / total * 100.0;
-        let confirm_failures_pct = tx_confirm_failures as f64 / total * 100.0;
-
-        info!(
-            "Stats since start:
-Runtime: {}s
-TPS: {:.4} tx/s ({:.4} symbols/s)
-Total attestation attempts: {}
-Successful txs: {} ({:.2})%
-Sending failures: {} ({:.2})%
-Confirmation failures: {} ({:.2})%
-Average batch resend delay: {:.2}s",
-            stats_start_time.elapsed().as_secs(),
-            tps,
-            sym_ps,
-            total,
-            tx_successes,
-            successes_pct,
-            tx_send_failures,
-            send_failures_pct,
-            tx_confirm_failures,
-            confirm_failures_pct,
-            batch_wait_times.as_secs() as f64 / (tx_successes as f64 + 0.000001),
+/// A future that decides how a batch is sent.
+///
+/// In daemon mode, attestations of the batch are scheduled
+/// continuously using spawn(), which means that a next attestation
+/// does not have to wait until the previous completes. Subsequent
+/// attestations are made according to the attestation_conditions
+/// field on the batch.
+///
+/// With daemon_mode off, this future attempts only one blocking
+/// attestation of the batch and returns the result.
+async fn attestation_sched_job(
+    mut batch: BatchState<'_>,
+    batch_no: usize,
+    batch_count: usize,
+    n_retries: usize,
+    retry_interval: Duration,
+    daemon: bool,
+    rpc_cfg: Arc<RLMutex<RpcCfg>>,
+    p2w_addr: Pubkey,
+    config: Pyth2WormholeConfig,
+    payer: Keypair,
+) -> Result<(), ErrBoxSend> {
+    let mut retries_left = n_retries;
+    // Enforces the max batch job count
+    let sema = Arc::new(Semaphore::new(batch.conditions.max_batch_jobs));
+    loop {
+        debug!(
+            "Batch {}/{}, group {:?}: Scheduling attestation job",
+            batch_no, batch_count, batch.group_name
         );
-    }
 
-    let mut errors = Vec::new();
+        let job = attestation_job(
+            rpc_cfg.clone(),
+            batch_no,
+            batch_count,
+            batch.group_name.clone(),
+            p2w_addr,
+            config.clone(),
+            Keypair::from_bytes(&payer.to_bytes()).unwrap(), // Keypair has no clone
+            batch.symbols.to_vec(),
+            sema.clone(),
+        );
 
-    // Filter out errors
-    for (batch_no, state) in batches {
-        match state.get_status() {
-            Success { .. } => {}
-            FailedSend { last_err, .. } | FailedConfirm { last_err, .. } => {
-                errors.push(last_err.to_string())
+        if daemon {
+            // park this routine until a resend condition is met
+            loop {
+                if let Some(reason) = batch
+                    .should_resend(&lock_and_make_rpc(&rpc_cfg).await)
```

Comment:
This will create a new RPC client, and so a new connection, every time, which could potentially be problematic. Instead could you create one RPC client at the top of `attestation_sched_job` and then sleep inside this loop to rate-limit the calls?